codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res

```diff
@@ -186,89 +214,146 @@ let setChainFetcher = (self: t, chainFetcher: ChainFetcher.t) => {
   }
 }
 
-let getFirstArbitraryEventsItemForChain = (queue: array<Types.eventBatchQueueItem>, ~chain) =>
+let getFirstArbitraryEventsItemForChain = (
+  queue: array<Types.eventBatchQueueItem>,
+  ~chain,
+  ~fetchStatesMap: ChainMap.t<fetchStateWithData>,
+): option<isInReorgThresholdRes<FetchState.itemWithPopFn>> =>
   queue
   ->Utils.Array.findReverseWithIndex((item: Types.eventBatchQueueItem) => {
     item.chain == chain
   })
   ->Option.map(((item, index)) => {
-    FetchState.item,
-    popItemOffQueue: () => queue->Utils.Array.spliceInPlace(~pos=index, ~remove=1)->ignore,
+    let {heighestBlockBelowThreshold} = fetchStatesMap->ChainMap.get(item.chain)
+    let isInReorgThreshold = item.blockNumber > heighestBlockBelowThreshold
+    {
+      val: {
+        FetchState.item,
+        popItemOffQueue: () => queue->Utils.Array.spliceInPlace(~pos=index, ~remove=1)->ignore,
+      },
+      isInReorgThreshold,
+    }
   })
 
-let getFirstArbitraryEventsItem = (queue: array<Types.eventBatchQueueItem>) =>
+let getFirstArbitraryEventsItem = (
+  queue: array<Types.eventBatchQueueItem>,
+  ~fetchStatesMap: ChainMap.t<fetchStateWithData>,
+): option<isInReorgThresholdRes<FetchState.itemWithPopFn>> =>
   queue
   ->Utils.Array.last
-  ->Option.map(item => {FetchState.item, popItemOffQueue: () => queue->Js.Array2.pop->ignore})
+  ->Option.map(item => {
+    let {heighestBlockBelowThreshold} = fetchStatesMap->ChainMap.get(item.chain)
+    let isInReorgThreshold = item.blockNumber > heighestBlockBelowThreshold
+    {
+      val: {FetchState.item, popItemOffQueue: () => queue->Js.Array2.pop->ignore},
+      isInReorgThreshold,
+    }
+  })
 
 let popBatchItem = (
-  ~fetchStatesMap: ChainMap.t<PartitionedFetchState.t>,
+  ~fetchStatesMap: ChainMap.t<fetchStateWithData>,
   ~arbitraryEventQueue: array<Types.eventBatchQueueItem>,
   ~isUnorderedMultichainMode,
-): option<FetchState.itemWithPopFn> => {
+): isInReorgThresholdRes<option<FetchState.itemWithPopFn>> => {
   //Compare the peeked items and determine the next item
   switch fetchStatesMap->determineNextEvent(~isUnorderedMultichainMode) {
-  | Ok({chain, earliestEvent}) =>
+  | Ok({val: {chain, earliestEvent}, isInReorgThreshold}) =>
     let maybeArbItem = if isUnorderedMultichainMode {
-      arbitraryEventQueue->getFirstArbitraryEventsItemForChain(~chain)
+      arbitraryEventQueue->getFirstArbitraryEventsItemForChain(~chain, ~fetchStatesMap)
     } else {
-      arbitraryEventQueue->getFirstArbitraryEventsItem
+      arbitraryEventQueue->getFirstArbitraryEventsItem(~fetchStatesMap)
     }
     switch maybeArbItem {
     //If there is item on the arbitray events queue, and it is earlier than
     //than the earlist event, take the item off from there
-    | Some(itemWithPopFn)
+    | Some({val: itemWithPopFn, isInReorgThreshold})
```

Comment:
Can we have the `isInReorgThreshold` as part of the `eventBatchQueueItem` instead of calculating it in the chainManager? I think it'd simplify quite a lot of code. Honestly, I don't like the increased complexity, and I hope it can be simplified to a single bool field with primitive logic.

Reply:
It can't just be on the eventBatchQueue item because it's not a static value, it changes as the head moves on and it's only relevant at the point in time when we create a batch.

1. We would still have to reduce through the created batch to see whether any item was within the threshold
2. A queue with no item that was fetching in the threshold would also be a valid point to start saving items.